### ISSUE TITLE
[BUG] npm_package_from not set in lifecycle

### ISSUE THREAD

**Issue by chee** [npm/cli#2811]
https://github.com/npm/rfcs/blob/d2f8dc3cbf35217a31d1bb5491154b18d4fb2276/implemented/0021-reduce-lifecycle-script-environment.md

the rfc that got rid of most of the environment variables states there will be an `npm_package_from` variable in when a package is being installed. there isn't. `npm_package_resolved`, and `npm_package_integrity` are there.

`npm_package_from` is not.

## current behaviour
npm_package_from not set in `install` lifecycle hook

## expected behaviour
npm_package_from would be set in `install` lifecycle hook

## steps
- `npm install @chee/yeet 2>&1 | grep npm_package`
- look at the output, and note there is no `npm_package_from`
the `install` script of that package is `env && false`

## Environment:

- os: arch linux version 200 and something
- node: 15.10.0
- npm: 7.6.0

**Comment by isaacs** [comment]
I think we need to update that RFC to remove `npm_package_from`, to be honest.

We can't guarantee that `npm_package_from` is valid or consistent, since dependencies may be included in the tree to satisfy multiple different dependencies.

For example, consider this dependency graph:

```
project -> (a@1, b@1, c@github:user/c#branch-v1.x, d@github:user/d#semver:1.2)
a -> d@1
b -> d@1.2
c -> d@github:user/d#branch-v1.x
```

Let's say that the `github:user/d` repository has a branch named `branch-v1.x` where the tip of tree is the tagged release `v1.2.3`.  This satisfies all 4 dependencies!  So, when installing d, which of these should the `npm_package_from` be set to?

* `1` (from a)
* `1.2` (from b)
* `github:user/d#branch-v1.x` (from c)
* `github:user/d#semver:1.2` (from the root project)

If we wanted to arbitrarily pick one and set the environment variable to that, we have a few options:

1. The first dependency that causes the node to _attempt_ to be installed.  This is going to be non-deterministic, since dependencies are fetched in parallel (even though they are placed within the tree in a deterministic order).  So that's not a good approach.
2. Arbitrarily pick one of the edges once the node is in the ideal tree.  Should it be the shallowest node that depends on it?  The deepest?  The first alphabetically?  Anything we choose is going to be slightly lying, but at least it won't be nondeterministic.
3. Just don't bother.  (That's the status quo if we do nothing, hence this issue.)

It might help if you could share why you need this?

**Comment by ljharb** [comment]
fwiw I agree; this seems just as unreliable as `module.parent`, for the same reasons.

**Comment by chee** [comment]
so that we can run a script when somebody says `npm install thing` but not when they `cd thing; npm install`

**Comment by isaacs** [comment]
> so that we can run a script when somebody says `npm install thing` but not when they `cd thing; npm install`

I see!  Ok, I don't think you need `npm_package_from` for this, you were just using it as an indicator that _something_ is pulling it in as a dep, and that's why it's there.

A simple way to do that today (mostly) is to look at the `INIT_CWD` environment variable and the current working directory.  If those are the same, it's `cd thing; npm i`.  If they're different, it's `npm i thing` (or possibly `npm i thing-that-depends-on-thing`).  It doesn't work if you do `cd thing/a/b/c; npm i`, because the INIT_CWD will be `/path/to/thing/a/b/c` but the current working directory will be `/path/to/thing`.

We could add an environment variable that just contains npm's local prefix (which may or may not be the INIT_CWD or PWD), so you could definitively compare them.  Would that satisfy your need?

**Comment by chee** [comment]
Yeah, our workaround is to use `[ "$PWD" == "$INIT_CWD" ]` which works just fine.

I'm fine with your first idea of updating the spec to remove it. It would be nice if this could be its own lifecycle, or environment variable, so that it's easier to discover and the code looks less m̶̲̠̭̺̠̳̙̫̲̪͂͑̉å̸̢̟͙͔̹̬̜̒̇̄̽͌̒̍͗̀́͒̚̚̚͝ģ̶̲͖̖̩͎̭̜͍̦̮̞͖̀i̷̛̛͎̝̩̮̜̬͖̲̫̭̗̠̊̒͛̂͋͊̔̾͂̕͘͝c̶̨̭͎̼̞͈̜̺͇͔͔̎̀̈̇̄̏̏̿͑͌́͌́͝k̴͎̰̻̙͔̘̤̥̾͑̽͆̈́͆̆̑̏̉̔̓̅̕͜͝͠ā̷̡̢̯̭̲̱̲͓̥͓͑̓̆̎̆̇̾l̸̩̣͔̜̳͙͈̮̙̲̯͂͛͆̂̚̕̕̚, but this is just fine.
